src/ast/passes/semantic_analyser.cpp

```diff
@@ -3531,35 +3570,72 @@ void SemanticAnalyser::accept_statements(StatementList &stmts)
   }
 }
 
-void SemanticAnalyser::update_key_type(const Map &map, const MapKey &new_key)
+void SemanticAnalyser::update_key_type(const Map &map,
+                                       const SizedType &new_key_type)
 {
   if (const auto &key = map_key_.find(map.ident); key != map_key_.end()) {
     bool valid = true;
-    if (key->second.args_.size() == new_key.args_.size()) {
-      for (size_t i = 0; i < key->second.args_.size(); i++) {
-        SizedType &key_type = key->second.args_[i];
-        const SizedType &new_key_type = new_key.args_[i];
-        if (key_type.IsStringTy() && new_key_type.IsStringTy()) {
-          key_type.SetSize(
-              std::max(key_type.GetSize(), new_key_type.GetSize()));
-        } else if (key_type != new_key_type) {
+    bool replace_key = false;
+    if (key->second.GetTy() == new_key_type.GetTy()) {
+      if (key->second.IsStringTy()) {
+        key->second.SetSize(
+            std::max(key->second.GetSize(), new_key_type.GetSize()));
+      } else if (key->second.IsTupleTy()) {
+        auto &keyFields = key->second.GetFields();
+        auto &newKeyFields = new_key_type.GetFields();
+        if (keyFields.size() != newKeyFields.size()) {
           valid = false;
-          break;
+        } else {
+          for (size_t i = 0; i < keyFields.size(); i++) {
+            SizedType &key_type = keyFields[i].type;
+            const SizedType &key_field_type = newKeyFields[i].type;
+            if (key_type.GetTy() != key_field_type.GetTy()) {
+              valid = false;
+              break;
+            } else if (key_type.IsStringTy()) {
+              // We have to replace the whole key type because of tuple
+              // alignment/padding
+              if (key_field_type.GetSize() > key_type.GetSize()) {
+                replace_key = true;
+              }
+            } else if (key_type.GetSize() != key_field_type.GetSize()) {
+              valid = false;
+              break;
+            }
```

Comment:
This type checking currently only checks two levels deep, but needs to be recursive to catch mismatched nested types.

I can break it with this script:
```
BEGIN
{
  $a = ((uint8)1,((uint8)123,(uint64)1234));
  $b = ((uint8)4,((uint64)1234,(uint8)123));
  @map[$a] = 1;
  @map[$b] = 2;
}
```

Reply:
Wrangling nested tuples and attempting to modify their nested expressions/types is very complicated and out of scope for this PR so I just added a restriction on using nested tuples as map keys - also added a test for it.

Reply:
The example above works now.